### ISSUE TITLE
[1.16.5] Incompatible with Timeless-and-Classics due to duplicated class

### ISSUE THREAD

**Issue by 3TUSK** [ValkyrienSkies/Valkyrien-Skies-2#448]
## Synopsis

Because both Valkyrien Skies and [Timeless and Classics][ref-1] repackaged jackson-core but different version, there is a chance that the game will crash if both are installed together.

[ref-1]: https://www.curseforge.com/minecraft/mc-mods/timeless-and-classic-guns-tac

## Steps to reproduce

To reliably reproduce this issue, we need to make TaC loaded **before** Valkyrien Skies.

  1. Install Valkyrien Skies, Timeless and Classics and this mod: [Trigger-Crash-ValkyrienSkies-TaC.jar.zip](https://github.com/ClumsyAlien/TimelessandClassics_Reforged/files/11241189/Trigger-Crash-ValkyrienSkies-TaC.jar.zip)
     - Due to GitHub restriction on supported file types, you will need to remove the `.zip` extension after donwloading.
     - If you have security concern over random jar on the internet, you can craft one on your own by following these steps:
       1. Make an empty class with `@Mod`
       2. Create `META-INF/mods.toml`, fill in minimal information (i.e. mod ID, description, author, etc)
       3. Declare dependency: this mod should load **after** Timeless and Classics, and **before** Valkyrien Skies. This is the key step.
       4. Build the jar
  2. Launch the game
  3. Create a new world or join an existing one
  4. Game crash

## Logs

<details>
<summary>
Click to view
</summary>

```
---- Minecraft Crash Report ----
// I'm sorry, Dave.

Time: 4/15/23 8:52 PM
Description: Exception in server tick loop

java.lang.NoSuchFieldError: EXACT_FLOATS
	at com.fasterxml.jackson.dataformat.cbor.CBORParser.<clinit>(CBORParser.java:73) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.dataformat.cbor.CBORParserBootstrapper.constructParser(CBORParserBootstrapper.java:93) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.dataformat.cbor.CBORFactory._createParser(CBORFactory.java:427) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.dataformat.cbor.CBORFactory.createParser(CBORFactory.java:343) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.dataformat.cbor.CBORFactory.createParser(CBORFactory.java:330) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.dataformat.cbor.CBORFactory.createParser(CBORFactory.java:27) ~[valkyrienskies:?] {re:classloading}
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3708) ~[tac:0.3.5.3-1.16.5] {re:classloading}
	at org.valkyrienskies.mod.common.ShipSavedData.func_76184_a(ShipSavedData.java:76) ~[valkyrienskies:?] {re:mixin,re:classloading}
	at net.minecraft.world.storage.DimensionSavedDataManager.func_223409_c(DimensionSavedDataManager.java:69) ~[?:?] {re:mixin,re:classloading}
	at net.minecraft.world.storage.DimensionSavedDataManager.func_215753_b(DimensionSavedDataManager.java:52) ~[?:?] {re:mixin,re:classloading}
	at net.minecraft.world.storage.DimensionSavedDataManager.func_215752_a(DimensionSavedDataManager.java:37) ~[?:?] {re:mixin,re:classloading}
	at net.minecraft.server.MinecraftServer.handler$zcg000$postCreateLevels(MinecraftServer.java:1786) ~[?:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:A}
	at net.minecraft.server.MinecraftServer.func_240787_a_(MinecraftServer.java:382) ~[?:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:A}
	at net.minecraft.server.MinecraftServer.func_240800_l__(MinecraftServer.java:308) ~[?:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:A}
	at net.minecraft.server.integrated.IntegratedServer.func_71197_b(IntegratedServer.java:60) ~[?:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_240802_v_(MinecraftServer.java:621) ~[?:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:A}
	at net.minecraft.server.MinecraftServer.func_240783_a_(MinecraftServer.java:232) ~[?:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:A}
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292] {}


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- System Details --
Details:
	Minecraft Version: 1.16.5
	Minecraft Version ID: 1.16.5
	Operating System: Mac OS X (x86_64) version 10.16
	Java Version: 1.8.0_292, AdoptOpenJDK
	Java VM Version: OpenJDK 64-Bit Server VM (mixed mode), AdoptOpenJDK
	Memory: 330164088 bytes (314 MB) / 909115392 bytes (867 MB) up to 954728448 bytes (910 MB)
	CPUs: 8
	JVM Flags: 2 total; -Xms512m -Xmx1024m
	ModLauncher: 8.1.3+8.1.3+main-8.1.x.c94d18ec
	ModLauncher launch target: fmlclient
	ModLauncher naming: srg
	ModLauncher services: 
		/mixin-0.8.4.jar mixin PLUGINSERVICE 
		/eventbus-4.0.0.jar eventbus PLUGINSERVICE 
		/forge-1.16.5-36.2.39-launcher.jar object_holder_definalize PLUGINSERVICE 
		/forge-1.16.5-36.2.39-launcher.jar runtime_enum_extender PLUGINSERVICE 
		/accesstransformers-3.0.1.jar accesstransformer PLUGINSERVICE 
		/forge-1.16.5-36.2.39-launcher.jar capability_inject_definalize PLUGINSERVICE 
		/forge-1.16.5-36.2.39-launcher.jar runtimedistcleaner PLUGINSERVICE 
		/mixin-0.8.4.jar mixin TRANSFORMATIONSERVICE 
		/forge-1.16.5-36.2.39-launcher.jar fml TRANSFORMATIONSERVICE 
	FML: 36.2
	Forge: net.minecraftforge:36.2.39
	FML Language Providers: 
		javafml@36.2
		minecraft@1
		kotlinforforge@1.17.0
	Mod List: 
		forge-1.16.5-36.2.39-client.jar                   |Minecraft                     |minecraft                     |1.16.5              |DONE      |Manifest: NOSIGNATURE
		obfuscate-0.6.3-1.16.5.jar                        |Obfuscate                     |obfuscate                     |0.6.3               |DONE      |Manifest: NOSIGNATURE
		forge-1.16.5-36.2.39-universal.jar                |Forge                         |forge                         |36.2.39             |DONE      |Manifest: 22:af:21:d8:19:82:7f:93:94:fe:2b:ac:b7:e4:41:57:68:39:87:b1:a7:5c:c6:44:f9:25:74:21:14:f5:0d:90
		tac-r-0.3.5.3-1.16.5.jar                          |Timeless and Classics Guns    |tac                           |0.3.5.3             |DONE      |Manifest: NOSIGNATURE
		Dummy.jar                                         |Dummy thing                   |dummy                         |0.0.0               |DONE      |Manifest: NOSIGNATURE
		valkyrienskies-116-2.0.0-alpha6.jar               |Valkyrien Skies 2             |valkyrienskies                |2.0.0-alpha6        |DONE      |Manifest: NOSIGNATURE
		curios-forge-1.16.5-4.1.0.0.jar                   |Curios API                    |curios                        |1.16.5-4.1.0.0      |DONE      |Manifest: NOSIGNATURE
		cloth-config-4.16.91-forge.jar                    |Cloth Config v4 API           |cloth-config                  |4.16.91             |DONE      |Manifest: NOSIGNATURE
	Crash Report UUID: 101253e1-7129-4fa5-b329-e1febb5f4d07
	Player Count: 0 / 8; []
	Data Packs: vanilla, mod:obfuscate, mod:valkyrienskies, mod:forge, mod:tac (incompatible), mod:curios, mod:cloth-config (incompatible)
	Type: Integrated Server (map_client.txt)
	Is Modded: Definitely; Client brand changed to 'forge'
```
</details>

## Analysis

tl;dr: Valkyrien Skies (or vs-core, to be exact) should relocate the packages of all dependent jars when building.

Timesless-and-Classics currently bundles `jackson-core` of an unknown, older version, while Valkyrien Skies bundles `jackson-core` 2.14.0. Both bundles the jar without relocation.
In `jackson-core` 2.14.0, [a new enum field `StreamReadCapability.class`][ref-2] is introduced. Judging by crash report, Valkyrien Skies is indirectly using this new field. 
However, as Timesless-and-Classics also bundles the same class, but of a different version, when the older one from Timesless-and-Classics is loaded first, then Valkyrien Skies will inevitably crash because the older one does not have that new enum field, which the newer class is compiled against. 

[ref-2]: https://github.com/FasterXML/jackson-core/commit/e655e46e61d20340924128b2bcdf9466600d1aea
[ref-3]: https://github.com/johnrengelman/shadow

## Possible Solutions

Ideally, this should be addressed by using Gradle plugins such as [shadow][ref-3]. It will handle the relocation properly; all you need to do is declare the relocating destination. 

However, given that it is actually [`vs-core` that uses jackson][ref-4], I am not sure where we should address this issue. Thus, this is an issue ticket for discussion, not a pull request. 

[ref-4]: https://github.com/ValkyrienSkies/vs-core/blob/7502f9ff2ac406c59148dc9dfc294489225ff16c/impl/build.gradle.kts#L30-L35

**Comment by 3TUSK** [comment]
Linked: ClumsyAlien/TimelessandClassics_Reforged#81

**Comment by 3TUSK** [comment]
> but would it be possible to load Valkyrien Skies first to avoid this crash?

Yes. In step 1.1.c `Declare dependency: this mod should load after Timeless and Classics, and before Valkyrien Skies`, you flip the order, i.e. `this mod should load after Valkyrien Skies, and before TaC`, and then you get a mod that workaround the issue.

However, it is still an workaround. Should there be third mod that packages a yet another version of `jackson-core` without relocation, you would need to mess with the loading order again in the same manner.

**Comment by ShadowFox88** [comment]
How would I force this without adding a bunch of unrelated mods?

**Comment by 3TUSK** [comment]
> How would I force this without adding a bunch of unrelated mods?

> In step 1.1.c Declare dependency: this mod should load after Timeless and Classics, and before Valkyrien Skies, you flip the order, i.e. this mod should load after Valkyrien Skies, and before TaC, and then you get a mod that workaround the issue.

**Comment by ShadowFox88** [comment]
Sorry I didn't see that I could edit the mods.toml file in your provided zip archive.

**Comment by DrKaltsit** [comment]
按我这个装
clockwork-forge-0.0.1-alpha350ee73315
create-1.18.2-0.5.1.f
framework-0.4.2-1.18.2
tac-a-0.3.10.2-1.18.2（在mcmod.cn装）
valkyrienskies-118-2.1.1-beta.4(或Beta6)

**Comment by thetheaplant** [comment]
Use TACZ
